diff -ruN a/fs/file.c b/fs/file.c
--- a/fs/file.c	2021-06-07 00:47:27.000000000 +0200
+++ b/fs/file.c	2021-06-07 08:12:28.011120443 +0200
@@ -788,6 +788,7 @@
 
 	return ret;
 }
+EXPORT_SYMBOL(close_fd_get_file);
 
 void do_close_on_exec(struct files_struct *files)
 {
diff -ruN a/kernel/fork.c b/kernel/fork.c
--- a/kernel/fork.c	2021-06-07 00:47:27.000000000 +0200
+++ b/kernel/fork.c	2021-06-07 08:04:09.183171631 +0200
@@ -1134,6 +1134,7 @@
 		schedule_work(&mm->async_put_work);
 	}
 }
+EXPORT_SYMBOL_GPL(mmput_async);
 #endif
 
 /**
diff -ruN a/kernel/sched/core.c b/kernel/sched/core.c
--- a/kernel/sched/core.c	2021-06-07 00:47:27.000000000 +0200
+++ b/kernel/sched/core.c	2021-06-07 08:05:19.344293269 +0200
@@ -5770,6 +5770,7 @@
 	return (nice_rlim <= task_rlimit(p, RLIMIT_NICE) ||
 		capable(CAP_SYS_NICE));
 }
+EXPORT_SYMBOL_GPL(can_nice);
 
 #ifdef __ARCH_WANT_SYS_NICE
 
diff -ruN a/kernel/task_work.c b/kernel/task_work.c
--- a/kernel/task_work.c	2021-06-07 00:47:27.000000000 +0200
+++ b/kernel/task_work.c	2021-06-07 08:06:31.113446458 +0200
@@ -60,6 +60,7 @@
 
 	return 0;
 }
+EXPORT_SYMBOL(task_work_add);
 
 /**
  * task_work_cancel_match - cancel a pending work added by task_work_add()
diff -ruN a/mm/memory.c b/mm/memory.c
--- a/mm/memory.c	2021-06-07 00:47:27.000000000 +0200
+++ b/mm/memory.c	2021-06-07 08:07:52.046741082 +0200
@@ -1548,6 +1548,7 @@
 	mmu_notifier_invalidate_range_end(&range);
 	tlb_finish_mmu(&tlb);
 }
+EXPORT_SYMBOL_GPL(zap_page_range);
 
 /**
  * zap_page_range_single - remove user pages in a given range
diff -ruN a/mm/shmem.c b/mm/shmem.c
--- a/mm/shmem.c	2021-06-07 00:47:27.000000000 +0200
+++ b/mm/shmem.c	2021-06-07 08:08:54.663739096 +0200
@@ -4232,6 +4232,7 @@
 
 	return 0;
 }
+EXPORT_SYMBOL_GPL(shmem_zero_setup);
 
 /**
  * shmem_read_mapping_page_gfp - read into page cache, using specified page allocation flags.
diff -ruN a/security/security.c b/security/security.c
--- a/security/security.c	2021-06-07 00:47:27.000000000 +0200
+++ b/security/security.c	2021-06-07 08:10:45.129493164 +0200
@@ -750,24 +750,28 @@
 {
 	return call_int_hook(binder_set_context_mgr, 0, mgr);
 }
+EXPORT_SYMBOL_GPL(security_binder_set_context_mgr);
 
 int security_binder_transaction(struct task_struct *from,
 				struct task_struct *to)
 {
 	return call_int_hook(binder_transaction, 0, from, to);
 }
+EXPORT_SYMBOL_GPL(security_binder_transaction);
 
 int security_binder_transfer_binder(struct task_struct *from,
 				    struct task_struct *to)
 {
 	return call_int_hook(binder_transfer_binder, 0, from, to);
 }
+EXPORT_SYMBOL_GPL(security_binder_transfer_binder);
 
 int security_binder_transfer_file(struct task_struct *from,
 				  struct task_struct *to, struct file *file)
 {
 	return call_int_hook(binder_transfer_file, 0, from, to, file);
 }
+EXPORT_SYMBOL_GPL(security_binder_transfer_file);
 
 int security_ptrace_access_check(struct task_struct *child, unsigned int mode)
 {
