--- a/drivers/gpu/drm/radeon/radeon_pm.c  2011-10-24 09:10:05.000000000 +0200
+++ b/drivers/gpu/drm/radeon/radeon_pm.c  2011-12-12 21:28:13.446675955 +0100
@@ -571,7 +571,8 @@
 
 	/* default to profile method */
 	rdev->pm.pm_method = PM_METHOD_PROFILE;
-	rdev->pm.profile = PM_PROFILE_DEFAULT;
+	/* rdev->pm.profile = PM_PROFILE_DEFAULT; */ // AG test 2011-12-12 PM_PROFILE_LOW"
+	rdev->pm.profile = PM_PROFILE_LOW;           // AG test 2011-12-12
 	rdev->pm.dynpm_state = DYNPM_STATE_DISABLED;
 	rdev->pm.dynpm_planned_action = DYNPM_ACTION_NONE;
 	rdev->pm.dynpm_can_upclock = true;


