diff -ruN a/fs/file.c b/fs/file.c
--- a/fs/file.c	2021-03-14 22:41:02.000000000 +0100
+++ b/fs/file.c	2021-03-15 14:44:45.017683378 +0100
@@ -775,6 +775,7 @@
 
 	return ret;
 }
+EXPORT_SYMBOL(close_fd_get_file);
 
 void do_close_on_exec(struct files_struct *files)
 {
diff -ruN a/kernel/fork.c b/kernel/fork.c
--- a/kernel/fork.c	2021-03-14 22:41:02.000000000 +0100
+++ b/kernel/fork.c	2021-03-15 14:45:27.765881696 +0100
@@ -704,6 +704,7 @@
 		schedule_work(&mm->async_put_work);
 	}
 }
+EXPORT_SYMBOL_GPL(mmput_async);
 
 static inline void free_signal_struct(struct signal_struct *sig)
 {
diff -ruN a/kernel/sched/core.c b/kernel/sched/core.c
--- a/kernel/sched/core.c	2021-03-14 22:41:02.000000000 +0100
+++ b/kernel/sched/core.c	2021-03-15 14:46:31.006175076 +0100
@@ -5765,6 +5765,7 @@
 	return (nice_rlim <= task_rlimit(p, RLIMIT_NICE) ||
 		capable(CAP_SYS_NICE));
 }
+EXPORT_SYMBOL_GPL(can_nice);
 
 #ifdef __ARCH_WANT_SYS_NICE
 
diff -ruN a/kernel/task_work.c b/kernel/task_work.c
--- a/kernel/task_work.c	2021-03-14 22:41:02.000000000 +0100
+++ b/kernel/task_work.c	2021-03-15 14:48:02.382598983 +0100
@@ -57,6 +57,7 @@
 
 	return 0;
 }
+EXPORT_SYMBOL(task_work_add);
 
 /**
  * task_work_cancel - cancel a pending work added by task_work_add()
diff -ruN a/mm/memory.c b/mm/memory.c
--- a/mm/memory.c	2021-03-14 22:41:02.000000000 +0100
+++ b/mm/memory.c	2021-03-15 14:49:28.174996982 +0100
@@ -1548,6 +1548,7 @@
 	mmu_notifier_invalidate_range_end(&range);
 	tlb_finish_mmu(&tlb);
 }
+EXPORT_SYMBOL_GPL(zap_page_range);
 
 /**
  * zap_page_range_single - remove user pages in a given range
diff -ruN a/mm/shmem.c b/mm/shmem.c
--- a/mm/shmem.c	2021-03-14 22:41:02.000000000 +0100
+++ b/mm/shmem.c	2021-03-15 14:50:43.183344940 +0100
@@ -4233,6 +4233,7 @@
 
 	return 0;
 }
+EXPORT_SYMBOL_GPL(shmem_zero_setup);
 
 /**
  * shmem_read_mapping_page_gfp - read into page cache, using specified page allocation flags.
diff -ruN a/security/security.c b/security/security.c
--- a/security/security.c	2021-03-14 22:41:02.000000000 +0100
+++ b/security/security.c	2021-03-15 14:53:20.876076467 +0100
@@ -727,24 +727,28 @@
 {
 	return call_int_hook(binder_set_context_mgr, 0, mgr);
 }
+EXPORT_SYMBOL_GPL(security_binder_set_context_mgr);
 
 int security_binder_transaction(struct task_struct *from,
 				struct task_struct *to)
 {
 	return call_int_hook(binder_transaction, 0, from, to);
 }
+EXPORT_SYMBOL_GPL(security_binder_transaction);
 
 int security_binder_transfer_binder(struct task_struct *from,
 				    struct task_struct *to)
 {
 	return call_int_hook(binder_transfer_binder, 0, from, to);
 }
+EXPORT_SYMBOL_GPL(security_binder_transfer_binder);
 
 int security_binder_transfer_file(struct task_struct *from,
 				  struct task_struct *to, struct file *file)
 {
 	return call_int_hook(binder_transfer_file, 0, from, to, file);
 }
+EXPORT_SYMBOL_GPL(security_binder_transfer_file);
 
 int security_ptrace_access_check(struct task_struct *child, unsigned int mode)
 {
